File: scripts/bench_pipe_jsonrpc_server.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import sys
from typing import Any


def _coerce_int(value: Any, default: int) -> int:
    if isinstance(value, bool):
        return default
    if isinstance(value, int):
        return value
    if isinstance(value, str):
        try:
            return int(value)
        except Exception:
            return default
    return default


def _payload(size: int, *, fill: str) -> str:
    if size <= 0:
        return ""
    return (fill * ((size // len(fill)) + 1))[:size]


def _write_message(message: dict[str, Any]) -> None:
    sys.stdout.write(json.dumps(message, separators=(",", ":")) + "\n")


def _handle_request(request: dict[str, Any]) -> None:
    request_id = request.get("id")
    method = str(request.get("method") or "")
    params_obj = request.get("params")
    params = params_obj if isinstance(params_obj, dict) else {}

    if method != "bench.echo":
        _write_message(
            {
                "jsonrpc": "2.0",
                "id": request_id,
                "error": {"code": -32601, "message": f"unknown method: {method}"},
            }
        )
        sys.stdout.flush()
        return

    response_bytes = max(0, _coerce_int(params.get("response_bytes"), 0))
    push_count = max(0, _coerce_int(params.get("push_count"), 0))
    push_bytes = max(0, _coerce_int(params.get("push_bytes"), 0))
    ordinal = max(0, _coerce_int(params.get("ordinal"), 0))

    for index in range(push_count):
        _write_message(
            {
                "jsonrpc": "2.0",
                "method": "bench.push",
                "params": {
                    "request_id": request_id,
                    "ordinal": ordinal,
                    "index": index,
                    "payload": _payload(push_bytes, fill="push_"),
                },
            }
        )

    _write_message(
        {
            "jsonrpc": "2.0",
            "id": request_id,
            "result": {
                "ok": True,
                "ordinal": ordinal,
                "payload": _payload(response_bytes, fill="resp_"),
            },
        }
    )
    sys.stdout.flush()


def main() -> int:
    for raw_line in sys.stdin:
        line = raw_line.strip()
        if not line:
            continue
        try:
            request = json.loads(line)
        except Exception:
            _write_message(
                {
                    "jsonrpc": "2.0",
                    "id": None,
                    "error": {"code": -32700, "message": "parse error"},
                }
            )
            sys.stdout.flush()
            continue
        if not isinstance(request, dict):
            _write_message(
                {
                    "jsonrpc": "2.0",
                    "id": None,
                    "error": {"code": -32600, "message": "invalid request"},
                }
            )
            sys.stdout.flush()
            continue
        _handle_request(request)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: native/fws_pipe_pump/src/bin/bench_pipe_jsonrpc_server.rs
use serde_json::{json, Value};
use std::io::{self, BufRead, BufWriter, Write};

fn coerce_i64(value: Option<&Value>, default: i64) -> i64 {
    match value {
        Some(Value::Number(number)) => number.as_i64().unwrap_or(default),
        Some(Value::String(text)) => text.parse::<i64>().unwrap_or(default),
        _ => default,
    }
}

fn payload(size: usize, fill: &str) -> String {
    if size == 0 {
        return String::new();
    }
    let repeat_count = (size / fill.len()) + 1;
    fill.repeat(repeat_count).chars().take(size).collect()
}

fn write_message<W: Write>(writer: &mut W, message: &Value) -> io::Result<()> {
    serde_json::to_writer(&mut *writer, message)?;
    writer.write_all(b"\n")
}

fn handle_request<W: Write>(writer: &mut W, request: Value) -> io::Result<()> {
    let request_id = request.get("id").cloned().unwrap_or(Value::Null);
    let method = request
        .get("method")
        .and_then(Value::as_str)
        .unwrap_or_default();
    let params = request.get("params").and_then(Value::as_object);

    if method != "bench.echo" {
        write_message(
            writer,
            &json!({
                "jsonrpc": "2.0",
                "id": request_id,
                "error": {"code": -32601, "message": format!("unknown method: {method}")},
            }),
        )?;
        writer.flush()?;
        return Ok(());
    }

    let response_bytes =
        coerce_i64(params.and_then(|p| p.get("response_bytes")), 0).max(0) as usize;
    let push_count = coerce_i64(params.and_then(|p| p.get("push_count")), 0).max(0) as usize;
    let push_bytes = coerce_i64(params.and_then(|p| p.get("push_bytes")), 0).max(0) as usize;
    let ordinal = coerce_i64(params.and_then(|p| p.get("ordinal")), 0).max(0);

    for index in 0..push_count {
        write_message(
            writer,
            &json!({
                "jsonrpc": "2.0",
                "method": "bench.push",
                "params": {
                    "request_id": request_id,
                    "ordinal": ordinal,
                    "index": index,
                    "payload": payload(push_bytes, "push_"),
                },
            }),
        )?;
    }

    write_message(
        writer,
        &json!({
            "jsonrpc": "2.0",
            "id": request_id,
            "result": {
                "ok": true,
                "ordinal": ordinal,
                "payload": payload(response_bytes, "resp_"),
            },
        }),
    )?;
    writer.flush()
}

fn main() -> io::Result<()> {
    let stdin = io::stdin();
    let mut writer = BufWriter::new(io::stdout().lock());

    for line_result in stdin.lock().lines() {
        let line = line_result?;
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        let request: Value = match serde_json::from_str::<Value>(trimmed) {
            Ok(value) if value.is_object() => value,
            Ok(_) => {
                write_message(
                    &mut writer,
                    &json!({
                        "jsonrpc": "2.0",
                        "id": null,
                        "error": {"code": -32600, "message": "invalid request"},
                    }),
                )?;
                writer.flush()?;
                continue;
            }
            Err(_) => {
                write_message(
                    &mut writer,
                    &json!({
                        "jsonrpc": "2.0",
                        "id": null,
                        "error": {"code": -32700, "message": "parse error"},
                    }),
                )?;
                writer.flush()?;
                continue;
            }
        };
        handle_request(&mut writer, request)?;
    }

    Ok(())
}
